### PR TITLE
Handshake-DC: add support for Int in Streams

### PR DESCRIPTION
**Pull request by luisacicolini**
This PR enables defining streams of int and bool for the formalization of DC and Handshake dialects from CIRCT

### CODE REVIEW

**Comment by tobiasgrosser** [comment]
Nice. I am happy to see this being merged if you feel this is sensible.

**Comment by luisacicolini** [comment]
Yes! Syncing with main broke the example but as soon as that's fixed I'm happy to merge

**Comment by tobiasgrosser** [comment]
Nice. 👍

**Comment by luisacicolini** [comment]
now ready for review (except for broken test), CC: @ymherklotz :)

File: SSA/Projects/CIRCT/DC.lean [hunk 1]
```diff
@@ -30,9 +33,10 @@ not consuming any tokens, until a message becomes available on the other stream
 Note that consuming `none`s is still allowed (and in fact neccessary to make progress).
 
 -/
-def branch (x c  : Stream) : Stream × Stream :=
 
-  corec₂ (β := Stream × Stream) (x, c)
+def branch (x : Stream α) (c : Stream Bool) : Stream α × Stream α :=
+
+  corec₂ (β := Stream α × Stream Bool) (x, c)
```
Comment (bollu) [inline SSA/Projects/CIRCT/DC.lean]:
```suggestion
def branch (x : Stream α) (c : Stream Bool) : Stream α × Stream α :=
  corec₂ (β := Stream α × Stream Bool) (x, c)
```

kill newline.

File: SSA/Projects/CIRCT/DC.lean [hunk 2]
```diff
@@ -119,29 +123,41 @@ def altMerge (x y : Stream) : Stream :=
...
+| branch (t : Ty2)
+| fst (t : Ty2)
+| snd (t : Ty2)
 deriving Inhabited, DecidableEq, Repr
 
 inductive Ty
-| Stream : Ty
-| Stream2 : Ty
+| stream (ty2 : Ty2) : Ty
+| stream2 (ty2 : Ty2) : Ty
```
Comment (bollu) [inline SSA/Projects/CIRCT/DC.lean]: Can I have a single sentence explanation for what `stream` and `stream2` are? 
```suggestion
| stream (ty2 : Ty2) : Ty -- A stream of values of type `ty2`.
| stream2 (ty2 : Ty2) : Ty -- A product of streams of values of type `ty2`.
```

File: SSA/Projects/CIRCT/DCExample.lean [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+def x : Stream Bool := ofList [some true, none, some false, some true, some false]
+def c : Stream Bool := ofList [some true, some false, none, some true]
+
+def test : Stream Bool :=
+  BranchEg1.denote (Valuation.ofPair c x)
+
+def remNone (lst : List (Option Bool)) : List (Option Bool) :=
+  lst.filter (fun | some x => true
+                  | none => false)
+
```
Comment (bollu) [inline SSA/Projects/CIRCT/DCExample.lean:42]: Can I have a `#guard_msgs in #eval ...` that runs the program and checks that it does something sensible?

**Comment by bollu** [comment]
Overall LGTM. I didn't take too close a look, but I trust you, and it seems like we're trying to figure out the precise modelling, so bikeshedding isn't worth it!

File: SSA/Projects/CIRCT/DCExample.lean (hunk 3 above)
Comment (luisacicolini) [reply SSA/Projects/CIRCT/DCExample.lean:42]: unfortunately this example is currently broken and excluded from all testing (it's the one that broke after updating Lean) - is it okay if I add it after we figure out what the exact issue is?

**Comment by luisacicolini** [comment]
thank you @bollu for the review! Indeed this model is not definitive and potentially subject to changes as we refine it

**Comment by tobiasgrosser** [comment]
Can we now press the merge button. I am afraid the upcoming lean updates will risk this code to become outdated again. As soon as its merged, I can look into why the example breaks.

**Comment by luisacicolini** [comment]
perfect, merging now!

**Comment by tobiasgrosser** [comment]
Nice. Thank you for the quick response.

**Pull request merged**